File: sdi3-14.Cli-REST/src/uo/sdi/clients/rest/menu/items/MainMenuItem.java
package uo.sdi.clients.rest.menu.items;

import java.util.ArrayList;
import java.util.List;

import uo.sdi.clients.rest.menu.MenuItem;
import uo.sdi.clients.rest.restService.RestService;

public abstract class MainMenuItem implements MenuItem {

	protected RestService restService;

	public MainMenuItem(RestService restService) {
		this.restService = restService;
	}
	
	@Override
	public List<MenuItem> execute() {
		
		executeInternal();

		List<MenuItem> items = new ArrayList<MenuItem>();
		
		items.add(new ListCategoriesMenuItem(restService));
		items.add(new ListTaskFromCategoryMenuItem(restService));
		items.add(new CreateTaskMenuItem(restService));
		items.add(new MarkTaskAsFinishedMenuItem(restService));
		items.add(new LogoutMenuItem());
		
		return items;
	}

	protected abstract void executeInternal();
}

File: sdi3-14.Cli-REST-CSharp/sdi3-14.Cli-REST-CSharp/menu/menuItems/MainMenuItem.cs
using sdi3_14.Cli_REST_CSharp.dto;
using sdi3_14.Cli_REST_CSharp.menu.menuItems;
using sdi3_14.Cli_REST_CSharp.restService;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace sdi3_14.Cli_REST_CSharp.menu
{
    public abstract class MainMenuItem : MenuItem
    {
        protected RestService restService;

        public MainMenuItem(RestService restService)
        {
            this.restService = restService;
        }

        IList<MenuItem> MenuItem.Execute()
        {

            executeInternal();

            List<MenuItem> items = new List<MenuItem>();

            items.Add(new ListCategoriesMenuItem(restService));
            items.Add(new ListTaskFromCategoryIdMenuItem(restService));
            items.Add(new CreateTaskMenuItem(restService));
            items.Add(new MarkTaskAsFinishedMenuItem(restService));
            items.Add(new LogoutMenuItem());
            
            return items;
        }

        public abstract String GetMenuName();

        protected abstract void executeInternal();
    }
}
